File: src/book.py
import os
import sys
from enum import Enum
from pathlib import Path


class BookFormat(Enum):
    EPUB = "epub"
    PDF = "pdf"

    @classmethod
    def from_path(cls, path: Path) -> "BookFormat":
        ext = path.suffix[1:].lower()
        if ext == "epub":
            return cls.EPUB
        if ext == "pdf":
            return cls.PDF
        raise ValueError(f"not a book file: {path}")

    def extension(self) -> str:
        return self.value


class FileKind(Enum):
    BOOK = "book"
    DATA = "data"


class BookInfo:
    def __init__(self, path: Path):
        self.path: Path = path
        self.id: str = path.name
        self.format: BookFormat = self._detect_format(path)

    def _detect_format(self, path: Path) -> BookFormat:
        for entry in path.iterdir():
            if entry.is_file():
                try:
                    return BookFormat.from_path(entry)
                except ValueError:
                    continue
        raise ValueError(f"Valid book file not found in: {path}")

    def get_file(self, kind: FileKind) -> Path:
        ext = self.format.extension() if kind == FileKind.BOOK else "dat"
        for entry in self.path.iterdir():
            if (
                entry.is_file()
                and entry.name.startswith(self.id)
                and entry.suffix.lower() == f".{ext}"
            ):
                return entry
        return self.path / f"{self.id}.{ext}"

    def file_name(self, kind: FileKind) -> str:
        if kind == FileKind.BOOK:
            return f"{self.id}.{self.format.extension()}"
        return self.get_file(kind).name


def library_path(user_idx: str) -> Path:
    if sys.platform == "darwin":
        home = Path(os.environ.get("HOME", "~")).expanduser()
        return (
            home
            / "Library"
            / "Application Support"
            / "Ridibooks"
            / "library"
            / f"_{user_idx}"
        )
    if sys.platform == "win32":
        appdata = Path(os.environ.get("APPDATA", ""))
        if not appdata or not appdata.exists():
            raise ValueError("APPDATA environment variable not found")
        return appdata / "Ridibooks" / "library" / f"_{user_idx}"
    raise NotImplementedError("library_path() not implemented for this OS")


def discover_books(path: Path) -> list[BookInfo]:
    if not path.exists():
        return []
    infos: list[BookInfo] = []
    for entry in path.iterdir():
        if entry.is_dir():
            try:
                infos.append(BookInfo(entry))
            except ValueError:
                continue
    return infos

File: src/book.ts
import { existsSync, readdirSync, statSync } from "fs";
import { basename, join } from "path";
import { homedir, platform } from "os";

export enum BookFormat {
  EPUB = "epub",
  PDF = "pdf",
}

export function bookFormatFromPath(path: string): BookFormat {
  const ext = path.substring(path.lastIndexOf(".") + 1).toLowerCase();
  if (ext === "epub") return BookFormat.EPUB;
  if (ext === "pdf") return BookFormat.PDF;
  throw new Error(`not a book file: ${path}`);
}

export function bookFormatExtension(format: BookFormat): string {
  return format.toString();
}

export enum FileKind {
  BOOK = "book",
  DATA = "data",
}

export class BookInfo {
  public readonly path: string;
  public readonly id: string;
  public readonly format: BookFormat;

  constructor(path: string) {
    this.path = path;
    this.id = basename(path);
    this.format = this.detectFormat(path);
  }

  private detectFormat(path: string): BookFormat {
    const entries = readdirSync(path);
    for (const entry of entries) {
      const fullPath = join(path, entry);
      if (statSync(fullPath).isFile()) {
        try {
          return bookFormatFromPath(fullPath);
        } catch {
          continue;
        }
      }
    }
    throw new Error(`Valid book file not found in: ${path}`);
  }

  getFile(kind: FileKind): string {
    const ext = kind === FileKind.BOOK ? bookFormatExtension(this.format) : "dat";
    const entries = readdirSync(this.path);
    for (const entry of entries) {
      const fullPath = join(this.path, entry);
      if (
        statSync(fullPath).isFile() &&
        entry.startsWith(this.id) &&
        entry.toLowerCase().endsWith(`.${ext}`)
      ) {
        return fullPath;
      }
    }
    return join(this.path, `${this.id}.${ext}`);
  }

  fileName(kind: FileKind): string {
    if (kind === FileKind.BOOK) {
      return `${this.id}.${bookFormatExtension(this.format)}`;
    }
    const filePath = this.getFile(kind);
    return basename(filePath);
  }
}

export function libraryPath(userIdx: string): string {
  const plat = platform();
  if (plat === "darwin") {
    const home = homedir();
    return join(
      home,
      "Library",
      "Application Support",
      "Ridibooks",
      "library",
      `_${userIdx}`
    );
  }
  if (plat === "win32") {
    const appdata = process.env.APPDATA;
    if (!appdata || !existsSync(appdata)) {
      throw new Error("APPDATA environment variable not found");
    }
    return join(appdata, "Ridibooks", "library", `_${userIdx}`);
  }
  throw new Error("library_path() not implemented for this OS");
}

export function discoverBooks(path: string): BookInfo[] {
  if (!existsSync(path)) {
    return [];
  }
  const infos: BookInfo[] = [];
  const entries = readdirSync(path);
  for (const entry of entries) {
    const fullPath = join(path, entry);
    if (statSync(fullPath).isDirectory()) {
      try {
        infos.push(new BookInfo(fullPath));
      } catch {
        continue;
      }
    }
  }
  return infos;
}
